Fix flaky ExistsQueryBuilderTests.testToQuery

atris opened the pull request:
Test was generating field patterns that matched raw.derived_keyword, which doesnt support exists queries. Fixed by replacing problematic patterns with TEXT_FIELD_NAME.

Fixes #18274

atris commented:
@owaiskazi19 @andrross Please review

sandeshkr419 commented on server/src/test/java/org/opensearch/index/query/ExistsQueryBuilderTests.java:
```diff
@@ -89,9 +89,14 @@ protected ExistsQueryBuilder doCreateTestQueryBuilder() {
         }
         // also sometimes test wildcard patterns
         if (randomBoolean()) {
-            if (randomBoolean()) {
```
I think the original intent of having 2 `randomBoolean()` was to reduce the probability of using a wild-card pattern to `0.25` which is implied by the comment `// also sometimes test wildcard patterns`. 

Logically equivalent logic will be:

```
if (randomBoolean() && randomBoolean()) {
....wild-card related logic....
} 
```

Now, honestly I don't know the reason behind having wild-card patters `sometimes` instead of just randomly 50% of the time, and don't have arguments to decide either way. However, let's change the behavior only if we intend to, else let's keep it the way it is.

atris replied:
Good catch. You're right about the 25% probability. I've restored the nested randomBoolean() structure to preserve the original behavior.

 I agree we should only change the probability if there's a specific reason to.

sandeshkr419 commented on server/src/test/java/org/opensearch/index/query/ExistsQueryBuilderTests.java:
```diff
@@ -89,9 +89,14 @@ protected ExistsQueryBuilder doCreateTestQueryBuilder() {
+        }
+        // Avoid patterns that would match the derived field "raw.derived_keyword"
+        // This includes "raw", "raw*", "raw.xxx", "r*", "ra*", etc.
+        if (fieldPattern.equals("raw")
```
Isn't the equality `(fieldPattern.equals("raw")` case redundant with `fieldPattern.startsWith("raw")`?

sandeshkr419 commented on server/src/test/java/org/opensearch/index/query/ExistsQueryBuilderTests.java:
```diff
@@ -89,9 +89,14 @@ protected ExistsQueryBuilder doCreateTestQueryBuilder() {
+        // This includes "raw", "raw*", "raw.xxx", "r*", "ra*", etc.
+        if (fieldPattern.equals("raw")
+            || fieldPattern.startsWith("raw")
+            || (fieldPattern.contains("*") && "raw.derived_keyword".startsWith(fieldPattern.replace("*", "")))) {
```
Can the pattern matching checks can be moved before the wild-card logic block, so that you don't have to worry about replacing `"*"` with `""`?

In case you match the restricted patterns, override early and then when you add `*` to `fieldPattern`, that should be safe to do so.

What do you think?

atris replied:
Agreed, fixed.

andrross commented on CHANGELOG.md:
```diff
@@ -99,6 +99,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Fixed
 - Fix flaky tests in CloseIndexIT by addressing cluster state synchronization issues ([#18878](https://github.com/opensearch-project/OpenSearch/issues/18878))
+- Fix flaky ExistsQueryBuilderTests.testToQuery ([#18274](https://github.com/opensearch-project/OpenSearch/issues/18274))
```
We generally don't add changelog entries for non-user facing things like test fixes. There is some guidance [here](https://github.com/opensearch-project/OpenSearch/blob/d9a927464ec62634d520bc32bbd34104c4f4c097/CONTRIBUTING.md#changelog).

atris replied:
Fixed

sandeshkr419 reviewed (changes requested):
Thanks @atris for taking time to fix this flaky test. 
I have a few comments to refine it more.

Also. if this was reproduce-able, can you please verify the same (before/after) by running the test in your IDE n-times.

atris commented:
> Thanks @atris for taking time to fix this flaky test. I have a few comments to refine it more.
> 
> Also. if this was reproduce-able, can you please verify the same (before/after) by running the test in your IDE n-times.

Thanks for the review @sandeshkr419. I've addressed all your comments - removed the redundant check, moved the pattern check before wildcard logic, and preserved the original 25% probability for wildcards.

  I ran the test 30+ times locally with random seeds including the two failing seeds from the issue (8C4D6928119E4426 and D373738A924E04A). All pass now. Before the fix, these seeds  failed with the UnsupportedOperationException.

jainankitk reviewed (approved):
Thanks @atris for fixing this flaky test!

atris commented:
Thank you for reviewing. @andrross @jainankitk @sandeshkr419 

If it looks good, can we please merge?
